### PR TITLE
Allow setting braille display specific settings in the GUI

### LINKED ISSUE #7452: Allow braille display specific settings in the GUI
In NVDA's voice settings, a synthesizer can implement specific settings, such as inflection, etc. For this, the VoiceSettingsSlider class is available for use. I propose generalizing this custom settings system to allow custom settings for braille displays as well. Possibilities include:

* Dot firmness
* Enable/disable braille input. This is desirable for the handy tech driver, since one of the handy tech devices uses the space keys for either scrolling or braille input
* Enable/disable active tactile control, also for Handy Tech devices

CC @BramD

### PR DESCRIPTION
**Pull request by LeonarddeR**
Though this pr is a fully working implementation, I"m open for any feedback and am willing to change things drastically if desired.

### Link to issue number:
Closes #7452

### Summary of the issue:
Speech synthesizers support changing synth specific properties (such as volume, inflection, rate boost) in the graphical user interface. Braille display drivers currently don't have this ability. This means that braille display specific settings, such as dot firmness and whether to enable braille input, can't be changed from the GUI.

### Description of how this pull request fixes the issue:
1. Introduce a new driverHandler with a Driver class that contains all the functionality that has to be available both for synthesizer and braille display drivers. In the future, vision enhancement providers (such as magnifier, screen curtain and focus highlight) will also inherit from this class. Support for driver specific settings is now part of this class.
2. Introduce a DriverSettingsMixin class for settings panels to support driver specific GUI settings. Most if not all of the synthesizer specific settings logic has been moved to this mixin.
3. Added braille input and ATC toggles to the Handy Tech display driver as an example implementation. The eurobraille and ALVA drivers are not yet changed for this.
4. Renamed the synthSettingsRing module to settingsRing. The global settings ring will still be a synthesizer settings ring. However, global plugins can now implement their own ring for custom drivers, such as future magnifiers.

### Testing performed:
Tested changing handy tech driver parameters through the gui. Also changed several synth parameters through the synth settings ring and gui, and checked that the gui successfully updated itself accordingly.

### Known issues with pull request:
None

### Change log entry:
* New features
    + For braille display drivers that support it, driver settings can now be changed from the braille settings category in NVDA's settings dialog. (#7452)

### CODE REVIEW

**Comment by leonardder** [comment]
Just to be clear, I don't at all expect this to be reviewed before 2018.2. Furthermore, I don't expect this to conflict with braille display auto detection in a major way.

@bramd: As I recall that you were particularly interested in this idea, I'd also welcome a quick review from you, if you have the time.

**Comment by dkager** [comment]
I think a braille input toggle should be added to all drivers that potentially support this (and then I wonder if it should actually be driver-specific). You could also allow setting the braille input table to "No input".

**Comment by leonardder** [comment]
The problem with your idea is that it requires some significant changes in all InputGesture derivatives, and I'm afraid it is currently not possible to test them all. I can do this for Hims, but than, we need to agree on what key names to use.

**Comment by dkager** [comment]
So we can't receive input and ignore it just before sending it to liblouis? I agree that this would be ugly, but it would be a universal setting that way.

The other solution for the HumanWare and ALVA displays is to just disable the 8 braille keys.

**Comment by leonardder** [comment]
I closed this by accidentally removing the branch. Reopening.

**Comment by feerrenrut** [comment]
The PR comment mentions that this is not complete, could you elaborate on what is remaining?
Also, can you please update the changes.t2t file.

**Comment by leonardder** [comment]
[**@feerrenrut**](https://github.com/feerrenrut) commented on [1 aug. 2018 03:39 CEST](https://github.com/nvaccess/nvda/pull/8214#issuecomment-409420671 "2018-08-01T01:39:32Z - Replied by Github Reply Comments"):
> The PR comment mentions that this is not complete, could you elaborate on what is remaining?

THis is actually no longer true, I updated the comment accordingly.

> Also, can you please update the changes.t2t file.

I don't think that we should aim to have this in 2018.3. Therefore if you agree, I'd rather wait until there's some more clarity about when this is ready to merge. Note that, though the amount of changes is really huge from a line count perspective, it is basically mainly moving things around, renaming synth to driver, etc.

**Comment by feerrenrut** [comment]
> Note that, though the amount of changes is really huge from a line count perspective, it is basically mainly moving things around.

Would you have time to rebase the changes so that moves and renames can be reviewed separately from the rest of the changes?

> I don't think that we should aim to have this in 2018.3

I have started reviewing this, and will leave my comments. I'll probably need to review it again.

**Comment by leonardder** [comment]
[**@feerrenrut**](https://github.com/feerrenrut) commented on [1 aug. 2018 08:00 CEST](https://github.com/nvaccess/nvda/pull/8214#issuecomment-409458812 "2018-08-01T06:00:32Z - Replied by Github Reply Comments"):
> Would you have time to rebase the changes so that moves and renames can be reviewed separately from the rest of the changes?

I just rebased on master.

File: source/settingsRing.py [hunk 1]
```diff
@@ -2,15 +2,16 @@
...
-	"""a numeric synth setting. Has functions to set, get, increase and decrease its value """
-	def __init__(self,synth,setting,min=0,max=100):
-		self.synth=synth
-		self.setting=setting
-		self.min=min
-		self.max=max
-		self.step = setting.normalStep if isinstance(setting,synthDriverHandler.NumericSynthSetting) else 1
+class DriverSetting(baseObject.AutoPropertyObject):
+	"""a numeric driver setting. Has functions to set, get, increase and decrease its value """
+	def __init__(self,driver,setting,min=0,max=100):
```
Comment (feerrenrut) [inline source/settingsRing.py]: min, max only really make sense for numerical derivatives of this class. I'd prefer it was moved to the specialisation.

**Comment by feerrenrut** [comment]
> I just rebased on master.

Ah woops, not quite what I meant. I should have been more clear. I was hoping that when I do the final review of this, I can do it one commit at a time. First look at the moved code all in one commit, then a commit with the rename, then all other changes in final (or several other commits). They do not have to be in that order, but being able to isolate each of those groups really makes the process easier.

**Review by feerrenrut — COMMENTED** [review]
Why is the settings ring limited to synth / braille driver settings? It might be a good moment to consider this, there have been issues in the past requesting that the options in the settings ring are configurable. I'm tempted to suggest that the settings ring changes be made separately from this PR. Being able to adjust braille driver settings from the NVDA braille settings dialog is a great start!

**Comment by leonardder** [comment]
[**@feerrenrut**](https://github.com/feerrenrut) commented on [1 aug. 2018 09:46 CEST](https://github.com/nvaccess/nvda/pull/8214#issuecomment-409482696 "2018-08-01T07:46:54Z - Replied by Github Reply Comments"):
> I was hoping that when I do the final review of this, I can do it one commit at a time. First look at the moved code all in one commit, then a commit with the rename, then all other changes in final (or several other commits). They do not have to be in that order, but being able to isolate each of those groups really makes the process easier.

I agree that this should have been handy, however I'm afraid that it will be very hard to accomplish this.

I have a backup of the old branch without the rebase on master, would it be helpful for you if I'd restore that? Probably not as it will create outdated review comments.

File: source/braille.py [hunk 2]
```diff
@@ -2364,6 +2368,36 @@ def _handleAck(self):
 		self._awaitingAck = False
 		_BgThread.queueApc(_BgThread.executor)
 
+	@classmethod
+	def DotFirmnessSetting(cls,defaultVal,minVal,maxVal):
```
[the remark this answers is not in the document]
Comment (leonardder) [reply source/braille.py]: I added and tested this functionality for the handy tech driver.

Í'm not sure what to do with models that don't support this functionality. Settings are defined on the driver class, and we don't know whether a model supports a setting when at the class level.

File: source/driverHandler.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+			if onlyChanged and getattr(self,s.name)==val:
+				continue
+			try:
+				setattr(self,s.name,val)
+			except UnsupportedConfigParameterError:
+				log.debugWarning("Unsupported setting %s; ignoring"%s.name, exc_info=True)
+				continue
+
+	@classmethod
+	def _paramToPercent(cls, current, min, max):
```
[the remark this answers is not in the document]
Comment (leonardder) [reply source/driverHandler.py]: The problem is that they are used within many synthesizer classes, moving them out of the SynthDriver or one of its parent classes would cause backwards incompatibility. Having said that, I can move them back to the SynthDriver class if you desire.

File: source/gui/nvdaControls.py [hunk 4]
```diff
@@ -37,3 +38,41 @@ def OnSetFocus(self, evt):
 		self.SetSelection(0, numChars)
 		evt.Skip()
 
+class EnhancedInputSlider(wx.Slider):
+	"""
+	A slider that supports additional key events, such as pageup/pageDown.
```
[the remark this answers is not in the document]
Comment (leonardder) [reply source/gui/nvdaControls.py]: It turns out that my assumptions where wrong. Page up and page down support is to move the slider in bigger steps, but this support was already there in wx.Slider. Honestly, I really don't know now what's the purpose of this class. 😖

File: source/globalCommands.py [hunk 5]
```diff
@@ -745,7 +745,7 @@ def script_reviewMode_previous(self,gesture):
 			speech.speakTextInfo(pos)
 		else:
 			# Translators: reported when there are no other available review modes for this object 
-			ui.reviewMessage(_("No previous review mode"))
+			ui.message(_("No previous review mode"))
```
Comment (bramd) [inline source/globalCommands.py]: Why is this changed? And why is this change included in this PR?

File: source/gui/nvdaControls.py [hunk 6]
```diff
@@ -42,6 +42,41 @@ def OnSetFocus(self, evt):
 		self.SetSelection(0, numChars)
 		evt.Skip()
 
+class EnhancedInputSlider(wx.Slider):
```
Comment (bramd) [inline source/gui/nvdaControls.py]: Is this class added or moved from elsewhere? What does it do exactly? Might need a docstring.

**Review by bramd — CHANGES_REQUESTED** [review]
Over all, this seems fine to me except a few comments I've left inline. It's however hard to review due to the renames and moving of methods.

File: source/brailleDisplayDrivers/handyTech.py [hunk 7]
```diff
@@ -203,20 +204,32 @@ def display(self, cells):
 
 class AtcMixin(object):
 	"""Support for displays with Active Tactile Control (ATC)"""
+	supportedSettings=(
+		# Translators: Name of the ATC (active tactile control) setting in braille settings.
+		# Active tactile control should be left untranslated.
+		BooleanDriverSetting("atc",_("Enable &Active Tactile Control")),
```
[the remark this answers is not in the document]
Comment (leonardder) [reply source/brailleDisplayDrivers/handyTech.py]: Makes sense. I commented it out. I'd like to keep the commented out version though, because it makes sense for documenting purposes. Feel free to suggest otherwise.

File: source/driverHandler.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+	configSpec="string(default=None)"
+
+	def __init__(self,name, displayNameWithAccelerator,
+		availableInSynthSettingsRing=False, displayName=None, useConfig=True):
+		"""
+		@param name: internal name of the setting
+		@type name: str
+		@param displayNameWithAccelerator: the localized string shown in voice or braille settings dialog
+		@type displayNameWithAccelerator: str
+		@param availableInSynthSettingsRing: Will this option be available in synthesizer settings ring?
```
[the remark this answers is not in the document]
Comment (leonardder) [reply source/driverHandler.py]: Hmm, not sure about this one. The classes in synthDriverHandler are now deprecated, should I also undo the deprecation notice? @eferrenrut, what do you think?

File: source/globalCommands.py (hunk 5 above)
Comment (leonardder) [reply source/globalCommands.py]: I simply have no idea, good catch. Reverted.

File: source/gui/nvdaControls.py (hunk 6 above)
Comment (leonardder) [reply source/gui/nvdaControls.py]: This is a rename from gui.settingsDialogs.VoiceSettingsSlider, and I really do not know why it even exists.

File: source/gui/settingsDialogs.py [hunk 9]
```diff
@@ -928,144 +928,193 @@ def onOk(self, evt):
...
-		self.panel=panel
-		super(StringSynthSettingChanger,self).__init__(setting)
+class StringDriverSettingChanger(DriverSettingChanger):
+	"""Same as L{DriverSettingChanger} but handles combobox events."""
+	def __init__(self,driver,setting,container):
+		self.container=container
+		super(StringDriverSettingChanger,self).__init__(driver,setting)
 
 	def __call__(self,evt):
+		# Quick workaround to deal with voice changes.
```
Comment (feerrenrut) [inline source/gui/settingsDialogs.py]: Since voice has special behavior, I would prefer to see a `VoiceDriverSettingChanger` subclass. Then the check for the `voice` setting name can be used to create this subclass.

File: source/gui/settingsDialogs.py [hunk 10]
```diff
@@ -928,144 +928,193 @@ def onOk(self, evt):
...
-		checkbox.SetValue(getattr(getSynth(),setting.name))
+			lambda evt: setattr(self.driver,setting.name,evt.IsChecked()))
+		checkbox.SetValue(getattr(self.driver,setting.name))
 		if self.lastControl:
 			checkbox.MoveAfterInTabOrder(self.lastControl)
 		self.lastControl=checkbox
 		return checkbox
 
+	def updateDriverSettings(self, changedSetting=None):
+		"""Creates, hides or updates existing GUI controls for all of supported settings."""
```
Comment (feerrenrut) [inline source/gui/settingsDialogs.py]: I expect there is a way we could avoid all this `isInstance` stuff. Probably not now, but it would be nice to see if there is some way to simplify this code.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/gui/settingsDialogs.py (hunk 9 above)
Comment (leonardder) [reply source/gui/settingsDialogs.py]: I agree that this is still hacky. However this doesn't only mean that we need a subclass of StringDriverSettingChanger. There is also makeSliderSettingControl that needs to know about the different functor for voice.

File: source/gui/settingsDialogs.py (hunk 10 above)
Comment (leonardder) [reply source/gui/settingsDialogs.py]: I guess we could move the setting makers to the driverHandler, i.e. make them methods on the driver settings. This also adds the ability for driver authors to create their own setting makers for specific settings, which would actually be very cool in scope of the gui of the vision framework. It would also solve the issue with the exception that is now applied to the voice setting.

**Comment by leonardder** [comment]
I've tested the handytech and eurobraille driver changes. I have no ALVA display here, but the change is very minimal and very similar to the eurobraille change.

**Comment by leonardder** [comment]
@feerrenrut: I believe this was almost ready for a merge. As this pr is pretty big and the Python 3 transition will most likely have a major impact on it, do you think it might help to have this merged beforehand? Note that I wrote #9064 with this pr in mind, i.e. it eases implementation of the gui of simple vision enhancement providers, such as the highlighter.

File: source/driverHandler.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+		@rtype: bool
+		"""
+		return False
+
+	def isSupported(self,settingName):
+		"""Checks whether given setting is supported by the driver.
+		@rtype: l{bool}
+		"""
+		for s in self.supportedSettings:
+			if s.name==settingName: return True
```
Comment (feerrenrut) [inline source/driverHandler.py]: How does this work with translations? We shouldn't be comparing translated strings.
Oh, I see there is also displayName. I would have prefered name be ID I think.
But then StringParameterInfo has ID and name where name is the displayable string.

File: source/driverHandler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/driverHandler.py]: Why can't you do the same as you already have done, import it for backwards compatibility?

File: source/driverHandler.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+
+	def _get_configSpec(self):
+		return "boolean(default={defaultVal})".format(defaultVal=str(self.defaultVal).lower())
+
+class UnsupportedConfigParameterError(NotImplementedError):
+	"""
+	Raised when changing or retrieving a driver setting that is unsupported for the connected device.
+	"""
+
+class StringParameterInfo(object):
```
Comment (feerrenrut) [inline source/driverHandler.py]: Why doesn't `StringParameterInfo` also inherit from `DriverSetting`?

File: source/braille.py [hunk 13]
```diff
@@ -1622,14 +1623,14 @@ def setDisplayByName(self, name, isFallback=False, detected=None):
 		elif not isFallback and not detected:
 			self._disableDetection()
 
+		firstLoad = not config.conf["braille"].isSet(name)
+		if firstLoad:
+			config.conf["braille"][name] = {}
```
[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/braille.py]: Is this resolved?

File: source/gui/nvdaControls.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/gui/nvdaControls.py]: This comment seems to have been removed, but the class is still there and still supports pageup/pagedown?

File: source/gui/settingsDialogs.py (hunk 10 above)
Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: I think this should be pulled out, right now it's hard to read. What about the driver provides a factory for constructing these. The common ones can be pulled out of here and provided for re-use. 

For now, leave this as it is, in the interests of getting this merged.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/gui/settingsDialogs.py [hunk 14]
```diff
@@ -981,144 +981,193 @@ def onOk(self, evt):
...
-			newValue = self.Max
-		elif key == wx.WXK_END:
-			newValue = self.Min
+			changeVoice(self.driver,getattr(self.container,"_%ss"%self.setting.name)[evt.GetSelection()].ID)
+			self.container.updateDriverSettings(changedSetting=self.setting.name)
 		else:
-			evt.Skip()
-			return
-		self.SetValue(newValue)
+			setattr(self.driver,self.setting.name,getattr(self.container,"_%ss"%self.setting.name)[evt.GetSelection()].ID)
```
[the remark this answers is not in the document]
Comment (leonardder) [reply source/gui/settingsDialogs.py]: Done, as well as some others

File: source/driverHandler.py (hunk 11 above)
Comment (leonardder) [reply source/driverHandler.py]: Mm, yes, this is confusing. I'd be happy to change name to id, that's not a big deal. however, I want to use self.id, not self.ID. So I would like to propose making things uniform here. How about self.id and self.displayName for both settings and parameters?

File: source/driverHandler.py (hunk 12 above)
Comment (leonardder) [reply source/driverHandler.py]: Because it's not a setting, it is a parameter for a setting. There's no accelerator either, and I don't think parameter names are translatable.

File: source/gui/settingsDialogs.py (hunk 10 above)
Comment (leonardder) [reply source/gui/settingsDialogs.py]: I agree

**Comment by leonardder** [comment]
I seem to be unable to reply on some of your inline comments, e.g. https://github.com/nvaccess/nvda/pull/8214#discussion_r282120665. Do you have any idea why that could be the case?

As for https://github.com/nvaccess/nvda/pull/8214#discussion_r282111373, I'm afraid I don't understand. What do you want me to import?

File: source/driverHandler.py (hunk 11 above)
Comment (feerrenrut) [reply source/driverHandler.py]: > How about self.id and self.displayName for both settings and parameters?

Yes, that is fine with me.

**Comment by feerrenrut** [comment]
>  seem to be unable to reply on some of your inline comments,

I think this is because they are on "outdated" lines. That's kind of annoying.

> I'm afraid I don't understand. What do you want me to import?

I was trying to continue a thread that started in https://github.com/nvaccess/nvda/pull/8214#discussion_r206757144 . These methods were moved from `synthDriverHandler` .  Rather than put them in `driverHandler` it actually looks like they don't need to be class variables. I think I got confused when I referred to the import, ignore that part.  Again, splitting this up might be work that could happen another time. Overall, I'm trying to reduce the size and complexity of modules in NVDA.

**Comment by feerrenrut** [comment]
> I've had another look at this project ans stumbled upon some issues

Could you first explain what these issues were?

**Comment by leonardder** [comment]
> > I've had another look at this project ans stumbled upon some issues
> 
> Could you first explain what these issues were?

Sure.

As for point 1, the issue popped up when using a handy tech display which supports the braille input setting. This driver contains a script to toggle braille input. However, when changing the state of the parameter using the script, the new state was never saved in the configuration. This is now solved by saving settings on terminating and on pre-saving configuration.

I also mentioned the issue related to config profile switches above. My second point wasn't strictly spoken because of an issue, but to avoid duplicated code, especially when the vision framework is also going to use the driverHandler.

**Comment by feerrenrut** [comment]
If you are happy with the testing on this, can you merge in master? I'll look to merge this tomorrow.

**Comment by leonardder** [comment]
Done. Thanks for the reminder!

**Review by feerrenrut — APPROVED** [review]
